unload_databricks_data_to_s3.py

```diff
@@ -9,6 +10,146 @@
+        CASE WHEN c1._change_type = 'update_preimage' THEN c1.data ELSE c2.data END AS previous_version
+      ) AS data,
+      'update' as _change_type,
+      c1._commit_timestamp,
+      c1._commit_version
+    FROM FormattedChanges c1
+    LEFT JOIN FormattedChanges c2 ON c1._commit_version = c2._commit_version AND c1._commit_timestamp = c2._commit_timestamp AND c1._change_type <> c2._change_type AND c1.data.{id_column} = c2.data.{id_column}
+    WHERE c1._change_type IN ('update_postimage', 'update_preimage') AND c2._change_type IN ('update_postimage', 'update_preimage')
+
+    UNION ALL
+
+    SELECT
+      STRUCT(
+        c1.data AS current_version,
+        null AS previous_version
+      ) as data,
+      c1._change_type,
+      c1._commit_timestamp,
+      c1._commit_version
+    FROM FormattedChanges c1
+    WHERE c1._change_type NOT IN ('update_postimage', 'update_preimage')
+  ),
+  DistinctData AS (
+    SELECT DISTINCT
+      data as data,
+      _change_type,
+      _commit_timestamp,
+      _commit_version
+    FROM Combined
+  )
+  SELECT
+    data,
+    _change_type as mutation_type,
+    '{mutation_row_type}' AS data_type,
+    UUID() as mutation_insert_id,
+    unix_timestamp(_commit_timestamp) * 1000 as mutation_time_ms
+  FROM DistinctData
+"""
+
+
+def normalize_sql_query(sql: str) -> str:
+    """
+    Normalize sql query to remove new lines, leading/trailing spaces, and replace multiple spaces with single space.
+    :param sql: sql query
+    :return: normalized sql query
+    """
+    return ' '.join(sql.split())
+
+
+def determine_first_table_name_in_sql(normalized_sql: str, table_names: set[str]) -> str:
+    """
+    Determine the first table name in the SQL query.
+    :param normalized_sql: SELECT SQL query without extra spaces and new lines
+    :param table_names: table names
+    :return: first table name in the SQL query
+    """
+    min_table_index: int = len(normalized_sql)
+    first_table_name: str = ""
+
+    for table_name in table_names:
+        table_index: int = normalized_sql.find(table_name)
+        if table_index != -1 and table_index < min_table_index:
+            min_table_index = table_index
+            first_table_name = table_name
+
+    return first_table_name
+
+
+def copy_and_inject_cdf_metadata_column_names(normalized_sql: str) -> str:
+    """
+    Creates a copy of the input SQL query with the CDF metadata column names injected into the SELECT clause.
+    :param normalized_sql: SELECT SQL query without extra spaces and new lines
+    :return: copy of the SQL query with CDF metadata column names injected into the SELECT clause
+    """
+    # Inject CDF metadata column names into the first SELECT clause
+    # Uses regex to replace the first occurrence of "select". The re.IGNORECASE flag is used to make the pattern
+    # case-insensitive
+    pattern = re.compile("select", re.IGNORECASE)
```

Comment:
yeah, it won't. That's why I don't like to do the manual under-the-hood manipulation on customer input: you need to gather all possible ways the input could be represented (which is hard and error-prone by itself) and then cover all of those edge cases (which is even harder)...

Moreover, "the first table name function" won't work either... And here comes the requirement for the customer to explicitly provide us with a table name that is supposed to be a delta table (if we start with supporting only one for now) and a requirement that this will be the main/outer-most table the selection is being made from (which by itself will be hard to validate)...

I'll try to think about how this could be resolved in the short term to cover all edge cases; however, I'm open to suggestions. If there won't be a good solution candidate found in the short term (before late April), I'll go with either:
1. Trying to cover known edge cases like this, but this won't guarantee robustness in general UX
2. Introducing table selection UI for DataBricks import and supporting CDF-based unload only through that view (which, in the case of DataBricks, is very limiting as CDF doesn't support views...)
3. Requiring a specific structure of query that will limit the support only to simple queries

cc @joez2 as this is a high-risk for mutation support with good UX for DataBricks import (for some reason I cannot mention @ajhorst in this repo, but would like to mention him as well)

Reply:
`determine_first_table_name_in_sql` is taking the list of all table names from the Databricks source and finding the first occurrence of one in the SQL string (i.e one with the lowest `find` index), right? This could be an opportunity to use SQL AST parsing. That still will be hard to cover 100% of all applicable cases, but could cover many more than a string search. Something like your option 3, but with a broader definition of 'simple queries'

Reply:
Yeah, I was starting with SQL parsing, but then I thought injection of 3 column names may be simple enough without requiring extra library to add as a dependency (which will bring extra maintenance burden to our team)😅 However, it looks like we really need some SQL parsing through tree-like structuring to explicitly select the first outer-most table/selection clause

Reply:
the proposed solution will truncate _commit_timestamp metadata (`unix_timestamp(_commit_timestamp) * 1000 as mutation_time_ms`), which is required by Cargo to sort and process mutation actions in proper order

to guarantee that there will at most 1 mutation per Cargo batch, we then need to compress the mutation actions on our side by resolving any intermediate mutation state in a single batch

Reply:
hmm, yeah, `_commit_timestamp` is a big issue. There is another way - we can loop all rows in the dataFrame, create a dataframe && tempView for each row and apply SQL to that tempView created to get transformed data. Then we can add `_commit_timestamp` in the row back to that transformed data. Not sure how efficient and costly of creating the tempView for each row, we can test that out though. If the approach above doesn't work, then yeah felt like we will have to manipulate the SQL string to add `_commit_timestamp` column at least.

Reply:
> create a dataframe && tempView for each row 

for each row?😅 i doubt it will be scalable🙈

Reply:
another way (which is not the best in terms of UX) is to require `_commit_timestamp` and `_change_type` to be present in the output of the SQL (however, it's very error-prone as customer may add it in incorrect place, which will result in unexpected customer-driven behaviour, i.e. the same situation as with timestamp in time-based import)

Reply:
> for each row?😅 i doubt it will be scalable🙈
yeah, that's my concern as well, but not sure if spark will optimize the execution plan or not to not actually do the work per work, but that's my hope. 

> another way (which is not the best in terms of UX) is to require _commit_timestamp and _change_type to be present in the output of the SQL
yeah, agreed. Also putting `_commit_timestamp` is kind of confusing cause customers are not ingestion `_commit_timestamp` as an event property, and that's mostly used by us only.